runtimes/pythonrt/pythonrt.go

```diff
@@ -0,0 +1,453 @@
+	if dirExists(venvPath) {
+		return nil
+	}
+
+	log.Info("creating venv", zap.String("path", venvPath))
+	return createVEnv(pyExe, venvPath)
+}
+
+func (*pySVC) Get() sdktypes.Runtime { return Runtime.Desc }
+
+const archiveKey = "archive"
+
+func (py *pySVC) Build(ctx context.Context, fs fs.FS, path string, values []sdktypes.Symbol) (sdktypes.BuildArtifact, error) {
+	py.log.Info("build")
+
+	data, err := createTar(fs)
+	if err != nil {
+		py.log.Error("create tar", zap.Error(err))
+		return sdktypes.InvalidBuildArtifact, err
+	}
+
+	var art sdktypes.BuildArtifact
+	art = art.WithCompiledData(
+		map[string][]byte{
+			archiveKey: data,
+		},
+	)
+
+	return art, nil
+}
+
+type PyMessage struct {
+	Type    string `json:"type"`
+	Name    string `json:"name"`
+	Payload []byte `json:"payload"`
+	Func    struct {
+		Name string            `json:"name"`
+		Args []string          `json:"args"`
+		Kw   map[string]string `json:"kw"`
+	} `json:"func"`
+}
+
+// All Python handler function get all event information.
+var pyModuleFunc = kittehs.Must1(sdktypes.ModuleFunctionFromProto(&sdktypes.ModuleFunctionPB{
+	Input: []*sdktypes.ModuleFunctionFieldPB{
+		{Name: "event_type"},
+		{Name: "event_id"},
+		{Name: "original_event_id"},
+		{Name: "integration_id"},
+		{Name: "data"},
+	},
+}))
+
+func entriesToValues(xid sdktypes.ExecutorID, entries []string) (map[string]sdktypes.Value, error) {
+	values := make(map[string]sdktypes.Value)
+	for _, name := range entries {
+		fn, err := sdktypes.NewFunctionValue(xid, name, nil, nil, pyModuleFunc)
+		if err != nil {
+			return nil, err
+		}
+		values[name] = fn
+	}
+
+	return values, nil
+}
+
+/*
+Flow (backend/internal/sessions/sessionworkflows/workflow.go)
+- Run should return a run
+- It should have mainPath in Values as FunctionValue
+- AK will call run.Call
+-
+*/
+
+func (py *pySVC) Run(
+	ctx context.Context,
+	runID sdktypes.RunID,
+	mainPath string,
+	compiled map[string][]byte,
+	values map[string]sdktypes.Value,
+	cbs *sdkservices.RunCallbacks,
+) (sdkservices.Run, error) {
+	py.log.Info("run", zap.String("id", runID.String()), zap.String("path", mainPath))
+
+	env, err := cbs.Load(ctx, runID, "env")
+	if err != nil {
+		return nil, fmt.Errorf("can't load env : %w", err)
+	}
+	envMap := kittehs.TransformMap(env, func(key string, value sdktypes.Value) (string, string) {
+		return key, value.GetString().Value()
+	})
+	py.log.Info("env", zap.Any("env", envMap))
+
+	tarData := compiled[archiveKey]
+	if tarData == nil {
+		return nil, fmt.Errorf("%q note found in compiled data", archiveKey)
+	}
+
+	ri, err := runPython(py.log, venvPy, tarData, mainPath, envMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cleanup in case we had errors
+	killPy := true
+	defer func() {
+		if killPy {
+			py.log.Error("killing Python", zap.Int("pid", ri.proc.Pid))
+			if err := ri.proc.Kill(); err != nil {
+				py.log.Warn("kill", zap.Int("pid", ri.proc.Pid), zap.Error(err))
+			}
+		}
+	}()
+
+	conn, err := ri.lis.Accept()
```

Comment:
don't you  need to do this before line 180?

Reply:
No, line 180 is the code the launches Python, if it fails there's no Python process to kill.